File: molecular/src/main/java/com/hartwig/actin/molecular/orange/util/AminoAcid.java
package com.hartwig.actin.molecular.orange.util;

import java.util.Map;

import com.google.common.collect.Maps;

import org.jetbrains.annotations.NotNull;

public final class AminoAcid {

    private static final Map<String, String> TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER = Maps.newHashMap();

    static {
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Ala", "A"); // Alanine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Cys", "C"); // Cysteine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Asp", "D"); // Aspartic Acid
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Glu", "E"); // Glutamic Acid
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Phe", "F"); // Phenylalanine

        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Gly", "G"); // Glycine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("His", "H"); // Histidine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Ile", "I"); // Isoleucine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Lys", "K"); // Lysine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Leu", "L"); // Leucine

        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Met", "M"); // Methionine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Asn", "N"); // Asparagine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Pro", "P"); // Proline
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Gln", "Q"); // Glutamine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Arg", "R"); // Arginine

        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Ser", "S"); // Serine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Thr", "T"); // Threonine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Val", "V"); // Valine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Trp", "W"); // Tryptophan
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.put("Tyr", "Y"); // Tyrosine
    }

    private AminoAcid() {
    }

    @NotNull
    public static String forceSingleLetterAminoAcids(@NotNull String impact) {
        String convertedImpact = impact;
        for (Map.Entry<String, String> mapping : TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER.entrySet()) {
            convertedImpact = convertedImpact.replaceAll(mapping.getKey(), mapping.getValue());
        }
        return convertedImpact;
    }
}

File: molecular/src/main/java/com/hartwig/actin/molecular/orange/interpretation/AminoAcid.kt
package com.hartwig.actin.molecular.orange.interpretation

import com.google.common.collect.Maps

internal object AminoAcid {
    private val TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER: MutableMap<String, String> = Maps.newHashMap()

    init {
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Ala"] = "A" // Alanine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Cys"] = "C" // Cysteine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Asp"] = "D" // Aspartic Acid
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Glu"] = "E" // Glutamic Acid
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Phe"] = "F" // Phenylalanine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Gly"] = "G" // Glycine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["His"] = "H" // Histidine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Ile"] = "I" // Isoleucine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Lys"] = "K" // Lysine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Leu"] = "L" // Leucine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Met"] = "M" // Methionine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Asn"] = "N" // Asparagine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Pro"] = "P" // Proline
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Gln"] = "Q" // Glutamine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Arg"] = "R" // Arginine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Ser"] = "S" // Serine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Thr"] = "T" // Threonine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Val"] = "V" // Valine
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Trp"] = "W" // Tryptophan
        TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER["Tyr"] = "Y" // Tyrosine
    }

    fun forceSingleLetterAminoAcids(impact: String): String {
        var convertedImpact: String = impact
        for ((key, value) in TRI_LETTER_AMINO_ACID_TO_SINGLE_LETTER) {
            convertedImpact = convertedImpact.replace(key.toRegex(), value)
        }
        return convertedImpact
    }
}
